04.ls/ls.rb

```diff
@@ -1,18 +1,49 @@
 #!/usr/bin/env ruby
 
-dir = Dir.glob '*'
-FILES = dir.count
-COLUMN = 3
-ROW = FILES.ceildiv(COLUMN)
-
-MAX_ROW = dir.max_by(&:length).length + 2
-
-index = 0
-1.upto(ROW) do |row| # 行ごとに繰り返し
-  COLUMN.times do # 列ごとに繰り返し
-    print dir[index].to_s.ljust(MAX_ROW, ' ') unless dir[index].nil?
-    index += ROW # 行数の倍数 == 配列から取得するindex変数の値
+require 'optparse'
+require 'etc'
+
+option = nil
+ARGV.options do |opt|
+  opt.on('-l') { option = '-l' }
+  opt.parse!
+end
+
+def chmod(mode)
+  mode.chars.map do |num|
+    %w[- --x -w- -wx r-- r-x rw- rwx][num.to_i]
+  end.join
+end
+
+def l_option(dir)
+  dir.each do |name|
+    info = File.stat(name)
+    mode = info.mode.to_s(8)[-4..]
+    chmod = chmod(mode)
+    u_name = Etc.getpwuid(info.uid).name
+    group = Etc.getgrgid(info.gid).name
+    puts "#{chmod}  #{info.nlink} #{u_name}  #{group}  #{info.size}  #{info.mtime.mon} #{info.mtime.day} #{info.mtime.hour}:#{info.mtime.min} #{name}"
+  end
+end
+
+def origin_ls_command(dir)
+  files = dir.count
+  column = 3
+
+  linage = files.ceildiv(column)
+
+  max_row = dir.max_by(&:length).length + 2
+
+  index = 0
+  1.upto(linage) do |row| # 行ごとに繰り返し
+    column.times do # 列ごとに繰り返し
+      print dir[index].to_s.ljust(max_row, ' ') unless dir[index].nil?
+      index += linage # 配列から取得するindex変数の値 == 行数の倍数
+    end
+    index = row
+    print "\n"
   end
-  index = row
-  print "\n"
 end
+
+dir = Dir.glob('*')
+option == '-l' ? l_option(dir) : origin_ls_command(dir)
```

Comment:
考え方としてはこれでOKです 👍 
あとはメソッド名を見直していただきたいです 🙏 それぞれ以下のような問題があると思っています。

- l_option: オプションが `-l` でなくなった場合に、このメソッド名も見直す必要が出てくる
- origin_ls_command: このプラクティスでたまたま先に実装しただけで、これがオリジナル、という訳ではない

Reply:
>l_option: オプションが -l でなくなった場合

メソッド名はオプションを指す命名から機能を指す命名に変更しました。
`origin_ls_command` も同様の考え方で修正しましたので、認識あっているか確認頂けないでしょうか。
```suggestion
option == '-l' ? list_detailed(dir) : list(dir)
```